OrbitCore/FunctionFramepointerValidator.cpp

```diff
@@ -0,0 +1,131 @@
+// Copyright (c) 2020 The Orbit Authors. All rights reserved.
+// Use of this source code is governed by a BSD-style license that can be
+// found in the LICENSE file.
+
+#include "FunctionFramepointerValidator.h"
+
+#include <capstone/capstone.h>
+#include <iostream>
+
+FunctionFramepointerValidator::FunctionFramepointerValidator(
+    csh handle, const uint8_t* code, size_t code_size) {
+  handle_ = handle;
+  instructions_ = nullptr;
+  instructions_count_ =
+      cs_disasm(handle, code, code_size, 0, 0, &instructions_);
+}
+
+FunctionFramepointerValidator::~FunctionFramepointerValidator() {
+  cs_free(instructions_, instructions_count_);
+}
+
+bool FunctionFramepointerValidator::isCallInstruction(
+    const cs_insn& instruction) {
+  for (uint8_t i = 0; i < instruction.detail->groups_count; i++) {
+    if (instruction.detail->groups[i] == X86_GRP_CALL) return true;
+  }
+  return false;
+}
+
+bool FunctionFramepointerValidator::isMovInstruction(
+    const cs_insn& instruction) {
+  return instruction.id == X86_INS_MOV || instruction.id == X86_INS_MOVQ;
+}
+
+bool FunctionFramepointerValidator::isBasePointer(uint16_t reg) {
+  return reg == X86_REG_BP || reg == X86_REG_EBP || reg == X86_REG_RBP;
+}
+
+bool FunctionFramepointerValidator::isStackPointer(uint16_t reg) {
+  return reg == X86_REG_SP || reg == X86_REG_ESP || reg == X86_REG_RSP;
+}
+
+// we will actually only check, whether there is a correct epilogue
+bool FunctionFramepointerValidator::validateEpilogue() {
+  bool found_epilogue = false;
+  for (size_t i = 0; i < instructions_count_ - 1; i++) {
+    if (isMovInstruction(instructions_[i]) &&
+        instructions_[i + 1].id == X86_INS_POP) {
+      cs_regs regs_read_first, regs_write_first, regs_read_second,
+          regs_write_second;
+      uint8_t read_count_first, write_count_first, read_count_second,
+          write_count_second;
+      // check for "mov esp, ebp; pop ebp"
+      if (  // first instruction
```

Comment:
This condition is unreadable. I would split it in a few boolean variables, *at least* one for "first instruction" and one for "second instruction", but possibly more.